empress/support_files/js/animator.js

```diff
@@ -0,0 +1,397 @@
+            animator.legendInfo[x] = { color: colorer.getColorHex(i) };
+        });
+
+        this.hide = hide;
+        this.lWidth = lWidth;
+    };
+
+    /**
+     * Clears state parameters
+     */
+    Animator.prototype.__resetParams = function() {
+        this.gradientCol = null;
+        this.trajectoryCol = null;
+        this.cm = null;
+        this.legendInfo = {};
+        this.gradientSteps = null;
+        this.totalFrames = null;
+        this.curFrame = null;
+        this.hide = null;
+        this.lWidth = null;
+        this.timePerFram = -1;
+        this.framesRdy = null;
+        this.queuedFrames = null;
+        this.pause = true;
+    };
+
+    /**
+     * Set thickness of colored branches
+     *
+     * @param {Number} lWidth How thick to make branches
+     */
+    Animator.prototype.setLineWidth = function(lWidth) {
+        this.lWidth = lWidth;
+    };
+
+    /**
+     * Set the hide status of uncolored branches
+     *
+     * @param {Boolean} hide If true, then Animator will hide all uncolored
+     *                  branches
+     */
+    Animator.prototype.setHide = function(hide) {
+        this.hide = hide;
+    };
+
+    /**
+     * Collects all the frames in the animation. This method is called while
+     * the asynchronous animation loop is running and is respondible for setting
+     * the timeframe ready flag to true once is has been collected.
+     *
+     * @private
+     */
+    Animator.prototype.__collectFrames = async function() {
+        for (var i = 0; i < this.totalFrames; i++) {
+            this.queuedFrames[i] = this.retriveFrame(i);
+            this.framesRdy[i] = true;
+        }
+    };
+
+    /**
+     * Draws the current frame and updates the legend.
+     */
+    Animator.prototype.drawFrame = function() {
+        var frame = this.queuedFrames[this.curFrame];
+        var name = frame.name;
+        var keyInfo = frame.keyInfo;
+        var obs = frame.obs;
+
+        // draw new legend
+        this.legend.clearAllLegends();
+        this.legend.addColorKey(name, keyInfo, "node", false);
+
+        //draw tree
+        this.empress.resetTree();
+        this.empress._colorTree(obs, this.cm);
+        this.empress.thickenSameSampleLines(this.lWidth);
+
+        // // TODO: hide should be taken care of in empress state machine
+        this.empress.setNonSampleBranchVisibility(this.hide);
+        this.empress.drawTree();
+        this.curFrame += 1;
+    };
+
+    /**
+     * The asynchronous animation loop. This method will continous create
+     * timeout events until animation is done or user pauses animation. If a
+     * frame is reached before it is ready, then a new timeout event will be
+     * created.
+     */
+    Animator.prototype.playAnimation = async function() {
+        // used in closure
+        var animator = this;
+
+        // Animation loop
+        var animationID = setTimeout(function loop() {
```

Comment:
Looks like the animationID is not used for anything, right?

```suggestion
        setTimeout(function loop() {
```

Reply:
Correct, this was an artifact left over from a previous method I was exploring.